### PR TITLE
authorization: add system:masters-only deep SAR via X-Kcp-Internal-Deep-SubjectAccessReview:true header

### PR DESCRIPTION
**Pull request by sttts**
This will allow us
- to authorize APIBindings to bind without actual APIExport workspace access
- to authorize use of CWTs without actual CWT workspace access (fixes https://github.com/kcp-dev/kcp/issues/1518)
- to authorize permission claim access to fall under MaximalPermissionClaims of the APIExport owner.

TODO:
- [x] ~~clarify whether we want to map the users somehow with a prefix~~ only local (SA) users would need a prefix, see next point.
- [ ] clarify whether we want to keep restricting service accounts to their workspace, and/or wipe their usernames when not having workspace access

### CODE REVIEW

File: pkg/authorization/deep_sar.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+// WithDeepSARConfig returns a clone of the input rest.Config
+// with an additional header making SARs to be deep.
+func WithDeepSARConfig(config *rest.Config) *rest.Config {
+	clone := *config
+	clone.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		return &withHeaderRoundtripper{
+			RoundTripper: rt,
+			headers: map[string]string{
+				deepSARHeader: "1",
```
Comment (sttts) [inline pkg/authorization/deep_sar.go:50]: "1" look strange. "true" probably is more common.

**Comment by sttts** [comment]
This will also help to get rid of the "bind to an ancestor or a child of an ancestor" rule. This is related to the API catalog discussion of binding against an APIExportKey cc @ncdc

File: pkg/authorization/deep_sar.go (hunk 1 above)
Comment (sttts) [reply pkg/authorization/deep_sar.go]: fixed

File: pkg/authorization/deep_sar.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+type withHeaderRoundtripper struct {
+	http.RoundTripper
+
+	headers map[string]string
+}
+
+func (rt *withHeaderRoundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req = req.Clone(req.Context())
```
Comment (s-urbaniak) [inline pkg/authorization/deep_sar.go:64]: why is this necessary? it looks like a noop.

Comment (s-urbaniak) [reply pkg/authorization/deep_sar.go:64]: nevermind:
```
	// RoundTrip should not modify the request, except for
	// consuming and closing the Request's Body. RoundTrip may
	// read fields of the request in a separate goroutine. Callers
	// should not mutate or reuse the request until the Response's
	// Body has been closed.
```

**Comment by s-urbaniak** [comment]
in favor of https://github.com/kcp-dev/kcp/pull/1739
